pkg/actors/health/health.go

```diff
@@ -0,0 +1,297 @@
+limitations under the License.
+*/
+
+package health
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"sync"
+	"time"
+
+	"k8s.io/utils/clock"
+
+	"github.com/dapr/kit/logger"
+)
+
+const (
+	defaultInitialDelay           = time.Second * 1
+	defaultFailureThreshold       = int32(2)
+	defaultRequestTimeout         = time.Second * 2
+	defaultHealthyStateInterval   = time.Second * 3
+	defaultUnHealthyStateInterval = time.Second / 2
+	defaultSuccessStatusCode      = http.StatusOK
+)
+
+var healthLogger = logger.NewLogger("actorshealth")
+
+type options struct {
+	client                 *http.Client
+	address                string
+	initialDelay           time.Duration
+	requestTimeout         time.Duration
+	failureThreshold       int32
+	healthyStateInterval   time.Duration
+	unhealthyStateInterval time.Duration
+	successStatusCode      int
+	clock                  clock.WithTicker
+}
+
+// Option is a function that applies a health check option.
+type Option func(o *options)
+
+// Checker is a health checker which reports to whether the target actor app
+// endpoint is in a healthy state, according to the configured status code.
+type Checker struct {
+	client *http.Client
+
+	ch chan bool
+
+	address                string
+	initialDelay           time.Duration
+	requestTimeout         time.Duration
+	failureThreshold       int32
+	healthyStateInterval   time.Duration
+	unhealthyStateInterval time.Duration
+	successStatusCode      int
+
+	clock   clock.WithTicker
+	closeCh chan struct{}
+	once    sync.Once
+	wg      sync.WaitGroup
+}
+
+func New(opts ...Option) (*Checker, error) {
+	options := &options{
+		failureThreshold:       defaultFailureThreshold,
+		initialDelay:           defaultInitialDelay,
+		requestTimeout:         defaultRequestTimeout,
+		successStatusCode:      defaultSuccessStatusCode,
+		healthyStateInterval:   defaultHealthyStateInterval,
+		unhealthyStateInterval: defaultUnHealthyStateInterval,
+		clock:                  new(clock.RealClock),
+	}
+
+	for _, o := range opts {
+		o(options)
+	}
+
+	if len(options.address) == 0 {
+		return nil, errors.New("required option 'address' is missing")
+	}
+
+	if options.client == nil {
+		options.client = new(http.Client)
+	}
+
+	return &Checker{
+		ch:                     make(chan bool),
+		client:                 options.client,
+		address:                options.address,
+		initialDelay:           options.initialDelay,
+		requestTimeout:         options.requestTimeout,
+		failureThreshold:       options.failureThreshold,
+		healthyStateInterval:   options.healthyStateInterval,
+		unhealthyStateInterval: options.unhealthyStateInterval,
+		successStatusCode:      options.successStatusCode,
+		clock:                  options.clock,
+		closeCh:                make(chan struct{}),
+	}, nil
+}
+
+func (c *Checker) Run(ctx context.Context) {
+	c.wg.Add(2)
+	defer c.Close()
+	defer c.wg.Done()
+
+	ctx, cancel := context.WithCancel(ctx)
```

Comment:
Let's keep each wg.Add next to the wg.Done so:

1. It's clearer when they are tied to each other
2. It saves errors when we make changes in the future (a similar pattern in pkg/actors has caused a bunch of issues over the last months)

```suggestion
	defer c.Close()
	c.wg.Add(1)
	defer c.wg.Done()

	ctx, cancel := context.WithCancel(ctx)
    c.wg.Add(1)
```

Reply:
Please, this HAS caused merge conflicts that were hard to debug

Reply:
This has been implemented https://github.com/dapr/dapr/pull/7022/files#diff-b51da868ff9b911968be455196e2898445e06852e44a7a89882b06e1734d36ceR114